### PR TITLE
doc/interoperability: new chapter and section on CycloneDX

### PR DESCRIPTION
**Pull request by ghost**
Add a section to the Nixpkgs manual describing a taxonomy for putting Nix info in CycloneDX SBOMs.

When this is merged and published it would be linked into the list of [official taxonomies](https://github.com/CycloneDX/cyclonedx-property-taxonomy).

The normal way to do this is to create a separate Github repo with an administrative working group, but that's a bunch of noise and overhead that we don't need. By putting this information into Nixpkgs we version it with the releases and we can oblige anything in Nixpkgs that deals with these SBOMs to use the taxonomy that is the manual.

This change adds a new *Interoperability Standards* chapter to the manual because I couldn't find an existing chapter that was appropriate. I don't have a strong opinion on how the manual should be structured.

I have experimented with using the SBOM JSON as an intermediary format that can be used to build a Nix package - [build-nim-sbom.nix](https://git.sr.ht/~ehmry/nim_lk/tree/buildSBOM/item/build-nim-sbom.nix).

If this works then ideally upstream projects that adopt SBOMs can have Nix information injected into them in an automated way that provides a reproducible Nix package, without requiring whoever is responsible for maintaining the SBOM be able to read and write Nix expressions.

This has implications for [Genealogos](https://github.com/tweag/genealogos) so it should be reviewed by its developers.

Pinging @NixOS/documentation-team.

Ref https://github.com/tweag/genealogos/issues/58

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/interoperability/cycloneds.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+### `nix:narinfo` {#sec-interop.cylonedx-narinfo}
+
+Narinfo properties describe component archives that may be available from binary caches.
+The `nix:narinfo` properties should be accompanied by a `nix:store_path` property within the same property list.
+
+| Property                  | Description |
+|---------------------------|-------------|
+| `nix:narinfo:store_path`  | Store path for the given store component. |
+| `nix:narinfo:url`         | URL path component. |
+| `nix:narinfo:nar_hash`    | Hash of the file system object part of the component when serialized as a Nix Archive. |
```
Comment (ehmry) [inline doc/interoperability/cycloneds.md:33]: I prefer to use camel-case rather than snake-case because it saves a few characters and it's common Nix style.

File: doc/interoperability/cycloneds.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+| `nix:narinfo:file_size`   | The size of the compressed component archive itself. |
+| `nix:narinfo:deriver`     | The path to the derivation from which this component is produced. |
+| `nix:narinfo:system`      | The hardware and software platform on which this component is produced. |
+| `nix:narinfo:sig`         | Signatures claiming that this component is what it claims to be. |
+| `nix:narinfo:ca`          | Content address of this store object's file system object, used to compute its store path. |
+| `nix:narinfo:references`  | A whitespace separated array of store paths that this component references. |
+
+### `nix:fod` {#sec-interop.cylonedx-fod}
+
+FOD properties describe a Fixed-Output-Derivation as a method and attributes passed to a function that implements the given method.
```
Comment (fricklerhandwerk) [inline doc/interoperability/cycloneds.md:46]:
```suggestion
FOD properties describe a fixed-output derivation as a method, and attributes passed to a function that implements the given method.
```

This is a bit hard to parse, but I think I understand what it means after reading the table. Maybe add a link to the FOD glossary entry here too.

File: doc/interoperability/cycloneds.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+The `nix:fod:method` property is required and must be accompanied by a `nix:store_path` property within the same property list.
+All other properties in this namespace are method-specific.
+
+| Property         | Description |
+|------------------|-------------|
+| `nix:fod:method` | Nixpkg function that produces this FOD. Required. Examples: `"fetchzip"`, `"fetchgit"` |
+| `nix:fod:name`   | Derivation name, present when method is `"fetchzip"` |
+| `nix:fod:ref`    | Git ref, present when method is `"fetchgit"` |
```
Comment (fricklerhandwerk) [inline doc/interoperability/cycloneds.md:54]: Please link to the Git man pages (not the manual!) on ref and rev, this is known to throw people off.

File: doc/interoperability/cycloneds.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+| `nix:fod:rev`    | Git rev, present when method is `"fetchgit"` |
+| `nix:fod:sha256` | FOD hash |
+| `nix:fod:url`    | URL to fetch |
+
+
+`nix:fod` properties may be extracted and evaluated to a derivation using code similar to the following:
```
Comment (fricklerhandwerk) [inline doc/interoperability/cycloneds.md:60]: Does such a function exist already?

**Review by fricklerhandwerk — COMMENTED** [review]
Approved documentation-wise. The writing is very clear, great work. Since this is likely to be read by complete Nix beginners but touches a rather advanced topic, please generously add links to definitions of terms

**Comment by drupol** [comment]
A very minor nit, shouldn't we rename the file `cycloneds.md` into `cyclonedx.md` ?

File: doc/interoperability/cycloneds.md (hunk 4 above)
Comment (ehmry) [reply doc/interoperability/cycloneds.md:60]: I have code for this out-of-tree, which is based on what I did for `buildNimPackage` for [reading lockfiles](https://github.com/NixOS/nixpkgs/blob/0e9703512aaa0dfa34f15e4931a2f7d58a47bcb1/pkgs/development/compilers/nim/build-nim-package.nix#L42). I was using the methods `fetchzip` and `git` but I like `fetchgit` for consistency.

**Comment by ehmry** [comment]
@fricklerhandwerk the text is good because I copied most of it from the Nix manual.

I was thinking the property names should follow [Store object info JSON format](https://nix.dev/manual/nix/2.22/protocols/json/store-object-info) as close as possible but the `nix:narinfo` names are already used by Genealogos, so I left them as is.

@drupol the `cycloneds.md` name was a typo, that has been fixed.

File: doc/interoperability/cyclonedx.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+|------------------|-------------|
+| `nix:fod:method` | Nixpkg function that produces this FOD. Required. Examples: `"fetchzip"`, `"fetchgit"` |
+| `nix:fod:name`   | Derivation name, present when method is `"fetchzip"` |
+| `nix:fod:ref`    | [Git ref](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefrefaref), present when method is `"fetchgit"` |
+| `nix:fod:rev`    | [Git rev](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefrevisionarevision), present when method is `"fetchgit"` |
+| `nix:fod:sha256` | FOD hash |
+| `nix:fod:url`    | URL to fetch |
+
+
+`nix:fod` properties may be extracted and evaluated to a derivation using code similar to the following:
```
Comment (fricklerhandwerk) [inline doc/interoperability/cyclonedx.md:63]:
```suggestion
`nix:fod` properties may be extracted and evaluated to a derivation using code similar to the following, assuming a fictitious function `filterPropertiesToAttrs`:
```

Just to make clear this is illustrative and won't actually work.

**Comment by ehmry** [comment]
I want to see this merged in some form so we can get the Nix namespace registered.

**Review by ErinvanderVeen — APPROVED** [review]

File: doc/interoperability/cyclonedx.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  methods = {
+    fetchzip =
+      { name, url, sha256, ... }:
+      pkgs.fetchzip {
+        inherit name url sha256;
+      };
+  };
+
+in methods.${fodProps.method} fodProps
+
```
Comment (fricklerhandwerk) [inline doc/interoperability/cyclonedx.md:79]:
```suggestion
```

**Comment by fricklerhandwerk** [comment]
Merging since Genealogos author @ErinvanderVeen approved.

**Comment by ehmry** [comment]
@fricklerhandwerk when does the manul on nixos.org get updated?

**Comment by fricklerhandwerk** [comment]
That should happen automatically, but I don't maintain that site, @thilobillerbeck or @garbas may now.

**Comment by ehmry** [comment]
Its in the unstable manual and linked in the official taxonomy list https://github.com/CycloneDX/cyclonedx-property-taxonomy
